test/distributed/org/apache/cassandra/distributed/test/ShortReadProtectionTest.java

```diff
@@ -18,46 +18,309 @@
+        return result;
+    }
+
+    @BeforeClass
+    public static void setupCluster() throws IOException
+    {
+        cluster = init(Cluster.build()
+                              .withNodes(NUM_NODES)
+                              .withConfig(config -> config.set("hinted_handoff_enabled", false))
+                              .start());
+    }
+
+    @AfterClass
+    public static void teardownCluster()
+    {
+        if (cluster != null)
+            cluster.close();
+    }
+
+    @Before
+    public void setupTester()
+    {
+        tester = new Tester(quorum, flush, paging, coordinator);
+    }
+
+    @After
+    public void teardownTester()
+    {
+        tester.dropTable();
+    }
+
+    /**
+     * Tests SRP for tables with no clustering columns and with a deleted row.
+     * <p>
+     * See CASSANDRA-13880.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13880()}.
+     */
+    @Test
+    public void testSkinnyTableWithoutLiveRows()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes("INSERT INTO %s (id) VALUES (0)")
+              .onlyNode1("DELETE FROM %s WHERE id = 0")
+              .assertRows("SELECT DISTINCT id FROM %s WHERE id = 0")
+              .assertRows("SELECT id FROM %s WHERE id = 0 LIMIT 1");
+    }
+
+    /**
+     * Tests SRP for tables with no clustering columns and with alternated live and deleted rows.
+     * <p>
+     * See CASSANDRA-13747.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13747()}.
+     */
+    @Test
+    public void testSkinnyTableWithLiveRows()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes(0, 10, i -> String.format("INSERT INTO %%s (id) VALUES (%d)", i)) // order is 5,1,8,0,2,4,7,6,9,3
+              .onlyNode1("DELETE FROM %s WHERE id IN (1, 0, 4, 6, 3)") // delete every other row
+              .assertRows("SELECT DISTINCT token(id), id FROM %s",
+                          row(token(5), 5), row(token(8), 8), row(token(2), 2), row(token(7), 7), row(token(9), 9));
+    }
+
+    /**
+     * Tests SRP for tables with no clustering columns and with complementary deleted rows.
+     * <p>
+     * See CASSANDRA-13595.
+     * <p>
+     * Replaces Python dtest {@code consistency_test.py:TestConsistency.test_13595()}.
+     */
+    @Test
+    public void testSkinnyTableWithComplementaryDeletions()
+    {
+        tester.createTable("CREATE TABLE %s (id int PRIMARY KEY)")
+              .allNodes(0, 10, i -> String.format("INSERT INTO %%s (id) VALUES (%d)", i)) // order is 5,1,8,0,2,4,7,6,9,3
+              .onlyNode1("DELETE FROM %s WHERE id IN (5, 8, 2, 7, 9)") // delete every other row
+              .onlyNode2("DELETE FROM %s WHERE id IN (1, 0, 4, 6)") // delete every other row but the last one
+              .assertRows("SELECT id FROM %s LIMIT 1", row(3))
```

Comment:
It isn't consistent, but I've observed this failing on the parameterization `[quorum=false flush=false paging=true coordinator=2]`:

```
java.lang.AssertionError: Expected: [[3]]
Actual: [[9]]
	at org.apache.cassandra.distributed.shared.AssertUtils.fail(AssertUtils.java:193)
	at org.apache.cassandra.distributed.shared.AssertUtils.assertTrue(AssertUtils.java:181)
	at org.apache.cassandra.distributed.shared.AssertUtils.assertRows(AssertUtils.java:70)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest$Tester.assertRows(ShortReadProtectionTest.java:505)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest$Tester.access$500(ShortReadProtectionTest.java:418)
	at org.apache.cassandra.distributed.test.ShortReadProtectionTest.testSkinnyTableWithComplementaryDeletions(ShortReadProtectionTest.java:174)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

I'm guessing it's because partition 9 isn't deleted from node 2, and node 2 as a coordinator somehow doesn't filter out partition 9 from node 1's deletion...interesting.

Reply:
Same as before, that's probably due to writes having the same timestamp.